Add fielddata and scripting support for byte-sized vectors

jdconrad opened the pull request:
This change adds support fielddata and subsequently scripting for `byte` vectors. This is a follow up to #90774 and completes the initial work for #89784.

Design decisions:

* I separated all of the scripting classes into one set for `byte` and one set for `float`. This allows for potential optimizations for each type without having to convert everything to `float` directly or do different things for different types through control flow in a tight loop. This also gives a template for future types where we add another set of classes that will handle it correctly, and follows the existing scripting fields api design. I really think we should consider having Lucene do a similar design update.

* I tried a different approach of using the existing `float` based classes in scripting and adding `byte` to them, and I do not think it is nearly as clear or lends itself to future optimizations/updates. Each additional type will create significantly more complexity than the approach in this PR. Changes can be seen on this branch (https://github.com/jdconrad/elasticsearch/tree/8bitfd).

* Similarity functions for the most part will remain unchanged for the user if they use the existing class bindings. The class bindings take in a `List<Number>` for the query vector as it is now. We can then decide how to both check boundaries and convert the query vector into an appropriate array type based on the the `element_type` of the corresponding vector field.

The length:
I know this is a ridiculously long PR so I'm open for ideas on how to break it into smaller parts. A ton of this is tests (~2200 lines of it) to test byte code in the same way we already test float code. I consider the majority of the changes to be in the following files:
* DenseVectorFieldMapper
* VectorDVLeafFieldData
* DenseVector
* ByteKnnDenseVector
* ByteBinaryDenseVector
* VectorScoreScriptUtils

jdconrad committed:
add byte vectors for scripting

jdconrad committed:
link byte vector fields to mapper field data

jdconrad committed:
spotless

jdconrad committed:
Merge branch 'main' into 8bitfd4

jdconrad committed:
update byte methods to accommdate normalized float vectors

jdconrad committed:
split dense vector functions into byte and float

jdconrad committed:
bit of clean up

jdconrad committed:
remove local tests

jdconrad committed:
more tests

jdconrad committed:
more tests

jdconrad committed:
more tests

jdconrad committed:
more tests

jdconrad committed:
more tests

jdconrad committed:
fix npe and more tests

jdconrad committed:
Merge branch 'main' into 8bitfd4

jdconrad committed:
add tests for byte vectors in scripts

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @jdconrad, I've created a changelog YAML for you.

jdconrad committed:
Update docs/changelog/91184.yaml

jdconrad committed:
fix style

jdconrad committed:
fix forbidden methods

jdconrad committed:
fix style

jdconrad commented:
@elasticsearchmachine run elasticsearch-ci/packaging-tests-windows-sample

jdconrad committed:
update test for byte

jdconrad commented on server/src/main/java/org/elasticsearch/script/VectorScoreScriptUtils.java:
```diff
@@ -46,31 +96,37 @@ public DenseVectorFunction(ScoreScript scoreScript, List<Number> queryVector, St
                 queryMagnitude += value * value;
             }
             queryMagnitude = Math.sqrt(queryMagnitude);
+            field.getElementType().checkVectorBounds(this.queryVector);
```
So, for float this would add NaN and Infinite checking which I think would cause downstream issues anyway as you mentioned in the prior PR? If not I can change this to only operate on byte vectors. For byte it's the same bounds checking we do for ann queries.

benwtrent replied:
> So, for float this would add NaN and Infinite checking which I think would cause downstream issues anyway as you mentioned in the prior PR? 

Ah, I thought `checkVectorBounds` was also checking the dimensions. No worries here then.

jdconrad commented on server/src/main/java/org/elasticsearch/script/VectorScoreScriptUtils.java:
```diff
@@ -18,24 +19,73 @@
+                queryMagnitude += value * value;
+            }
+            this.qvMagnitude = (float) Math.sqrt(queryMagnitude);
+            field.getElementType().checkVectorBounds(vector);
```
So the design here I think works pretty well because for the normal use case users already use the class bindings methods which cache the query vector as part of the script instance (which amounts to per segment). And since the query vector is a `List<Number>` we can just check it's actually a byte and then use cache a `byte[]`. This becomes a lot more complex and less performant if users want to operate on the `DenseVector` directly.

jdconrad commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/VectorDVLeafFieldData.java:
```diff
@@ -57,11 +62,20 @@ public DocValuesScriptFieldFactory getScriptFieldFactory(String name) {
                     values = null;
                 }
-                return new KnnDenseVectorDocValuesField(values, name, dims);
+                if (elementType == ElementType.BYTE) {
```
Good call. I will make the change for all of the cases that do this.

jdconrad commented on modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/185_knn_and_binary_byte_dv_fields_api.yml:
```diff
@@ -0,0 +1,882 @@
+---
+"empty dense vector throws for vector accesses":
+  - skip:
+      version: " - 8.1.99"
```
Oops! I'll take another pass through the tests. I admit I copied and updated a lot of these directly from the float tests for byte.

benwtrent replied:
@jdconrad just so you don't forget here :) I think BWC will blow up here. Mind if we turn on test-full-bwc? That way we get CI to run on all the bwc versions to catch that we are skipping correctly here.

jdconrad replied:
You are correct! I just updated the skips this morning. Is there something special I have to do to ensure full-bwc is run?

jdconrad commented on server/src/main/java/org/elasticsearch/script/VectorScoreScriptUtils.java:
```diff
@@ -79,11 +135,35 @@ public double l1norm() {
+
+        public L1Norm(ScoreScript scoreScript, List<Number> queryVector, String fieldName) {
+            DenseVectorDocValuesField field = (DenseVectorDocValuesField) scoreScript.field(fieldName);
+            if (field.getElementType() == DenseVectorFieldMapper.ElementType.BYTE) {
```
I'll change this for all of them.

jdconrad commented on server/src/main/java/org/elasticsearch/script/field/vectors/BinaryDenseVector.java:
```diff
@@ -37,11 +38,27 @@ public float[] getVector() {
+        ByteBuffer byteBuffer = wrap(docVector);
+
+        double dotProduct = 0;
+        for (float v : queryVector) {
+            dotProduct += byteBuffer.getFloat() * v;
+        }
```
I think we should probably have a high-bandwidth discussion on how some of the byte/float calculations should interact. 

In this case `BinaryDenseVector` is the `float` version of this class. I'd like to rename it in a follow up to `FloatBinaryDenseVector` which would match `ByteBinaryDenseVector`, but didn't want to add to the already really high number of lines in this PR.

I guess the question I have is should this just be an `UnsupportedOperationException`? With scripting it's a bit strange because an unusual case could have a query across multiple indices where a field is a `float` vector and the same field in a different index is a `byte` vector.`

jdconrad replied:
Updated do stop mixed operations between `byte[]` and `float[]` vectors.

jdconrad commented on server/src/main/java/org/elasticsearch/script/field/vectors/ByteBinaryDenseVector.java:
```diff
@@ -0,0 +1,217 @@
+        return result;
+    }
+
+    protected double dotProductNormalized(float[] normalizedQueryVector) {
+        float result = 0;
+        int i = 0;
+        int j = docVector.offset;
+        while (i < dims) {
+            result += docVector.bytes[j++] * normalizedQueryVector[i++];
+        }
+        return result;
```
Agreed on the trickiness of this. I would like to add this to a high-bandwidth discussion topic as well. I wasn't really sure what to do here exactly, but it really did not seem valuable to have a normalized `byte` vector.

jdconrad replied:
Another discussion item around this as a similar topic -- it makes sense to check the query vector for correct bounds in class bindings because it's only called once per segment and cached. But if a user calls one of the methods directly on a `DenseVector` as a `field('my-vector').get().dotProduct(new float[] {...})` then checking could be a large performance hit. Right now we just cast, so the user is responsible for giving a valid vector in this particular case.

jdconrad replied:
Updated do stop mixed operations between byte[] and float[] vectors.

jdconrad commented:
@benwtrent Thank you for the review! I definitely think we should have a discussion around where it makes sense to use/cast byte and float in a number of different places. The other thought I had is maybe we should just make the two types `float` and `int` since for any operation bytes are going to be up-cast to `int` in Java anyway. Then we end up with the class structure as `FloatBinaryDenseVector`, `FloatKnnDenseVector`, `IntBinaryDenseVector`, and `IntKnnDenseVector`. But maybe that's bad for potential future vectoriazation.

jtibshirani commented:
Thanks @jdconrad, this will be really useful. Several users have found non-indexed `dense_vector to be helpful and I'm sure they'll appreciate a space-saving option.

One question I had is if we should try to simplify and do everything in terms of `float[]` values for a first pass. For example, instead of using `VectorValues#binaryValue`, we'd just use `VectorValues#vectorValue` even when the encoding type is `byte`. On the search side, we'd always accept `float[]` just like we do for ANN search. This could simplify a lot, because we would avoid duplicate function implementations, and the parallel class hierarchy for byte types (like `ByteKnnDenseVector`). I'm not sure there's a huge performance advantage to working with `byte[]` directly ... and if there is we could test that through a benchmark and refactor later? What do you think?

jtibshirani reviewed (commented):
Thanks @jdconrad, this will be really useful. Several users have found non-indexed `dense_vector` to be helpful and I'm sure they'll appreciate a space-saving option.

One question I had is if we should try to simplify and do everything in terms of `float[]` values for a first pass. For example, instead of using `VectorValues#binaryValue`, we'd just use `VectorValues#vectorValue` even when the encoding type is `byte`. On the search side, we'd always accept `float[]` just like we do for ANN search. This could simplify a lot, because we would avoid duplicate function implementations, and the parallel class hierarchy for byte types (like `ByteKnnDenseVector`). I'm not sure there's a huge performance advantage to working with `byte[]` directly ... and if there is we could test that through a benchmark and refactor later? What do you think?

jdconrad commented:
@jtibshirani Your design suggestion is actually the initial approach I took. I think there is some trade-off between clarity of the type breakdown and a bit of performance (as a follow up I would like to add some micro benchmarks which I talked to @benwtrent a bit about), but as an initial approach it would definitely be a smaller core change. I think testing will be about the same. 

I would like to take a bit of time to switch this over your suggested approach and post that as a different PR for comparison, and we can go from there. What do you think?

jdconrad commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:
```diff
@@ -830,7 +832,7 @@ public void write(XContentBuilder b) throws IOException {
             ByteBuffer byteBuffer = ByteBuffer.wrap(ref.bytes, ref.offset, ref.length);
             for (int dim = 0; dim < dims; dim++) {
-                b.value(byteBuffer.getFloat());
+                b.value(elementType == ElementType.FLOAT ? byteBuffer.getFloat() : byteBuffer.get());
```
Yes, I will add this!

jdconrad commented:
I did a bit of performance testing locally and will add some micro benchmarks as a follow up. For non-indexed binary vectors, byte vectors see a significant performance improvement for dot product and cosine similarity when using the split-class approach. For indexed knn vectors, byte vectors actually take a hit on dot product, but improve on cosine similarity.

After speaking w/ @jtibshirani we think the performance improvement is worth the overhead of the split class approach.

jdconrad committed:
partial response to pr comments

jdconrad committed:
throw unsupportedoperationexception when types do not match in
float/byte vector methods

jdconrad committed:
Merge branch 'main' into 8bitfd4

benwtrent commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/VectorDVLeafFieldData.java:
```diff
@@ -57,10 +62,16 @@ public DocValuesScriptFieldFactory getScriptFieldFactory(String name) {
             } else {
                 BinaryDocValues values = DocValues.getBinary(reader, field);
-                return new BinaryDenseVectorDocValuesField(values, name, dims, indexVersion);
+                return switch (elementType) {
+                    case BYTE -> new ByteBinaryDenseVectorDocValuesField(values, name, elementType, dims);
+                    case FLOAT -> new BinaryDenseVectorDocValuesField(values, name, elementType, dims, indexVersion);
+                };
```
SO CLEAN! Dang, I love these switch expressions. Thank you Java for slowly modernizing!

jdconrad replied:
This is MUCH nicer.

jdconrad committed:
fix skip in tests for new feature

jdconrad commented on modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/165_dense_vector_byte_special_cases.yml:
```diff
@@ -0,0 +1,245 @@
+  - match: {hits.hits.0._id: "1"}
+  - match: {hits.hits.1._id: "2"}
+
+  # query vector of type float
```
Good question!

Okay, so I went and added unit tests in `VectorScoreScriptUtilsTests` to cover the class-bindings portion of this, and invalid byte vectors are screened by `checkVectorBounds`, so there can be no invalid vectors here. 

However, for non-class bindings this gets a bit weird to support arity-only type overloading in Painless. (Note that if a user is doing this they are already doing something more advanced and less efficient, so we recommend they use the class-bindings which can cache the query vector and normalize when it makes sense to do so.)

The following can be passed into one of the vector calc methods on a byte vector:
byte[] - known to be valid
float[] - will throw an UnsupportedOperationException now
List<Number> - no checks and will convert anything to a byte

I think for the List<Number> case it would be *very* expensive to add bounds checking here. The user is already in a advanced use case, so I would advocate against this check in this singular case.

jdconrad committed:
add more tests for unsupported operations

jdconrad committed:
Merge branch 'main' into 8bitfd4

jdconrad committed:
update style

jdconrad commented:
@jtibshirani @benwtrent I've updated the following based on feedback:

1. I made it so all operations between byte vectors and float vectors are unsupported. As an example if a user attempts to use a `byte[]` query vector for dot product with a `float[]` doc vector this will throw an `UnsupportedOperationException`. The one caveat to this is if a user passes in a `List<Number>`. We do not explicitly check the instance type of the `Number`. We instead just call `getFloat()` or `getByte()` when appropriate. I would advocate for this simply to reduce additional complexity and avoid performance issues.
2. I updated unit tests to account for the new unsupported methods.
3. I fixed the skips for yml tests to ensure bwc tests can pass.

jdconrad committed:
fix unit test

jdconrad committed:
Merge branch 'main' into 8bitfd4

benwtrent reviewed (approved):
Read through it again. I think this is good to go. Lots of good work here.

jdconrad commented:
@benwtrent Thank you again for the reviews! I'll commit as soon as I can get a clean CI run.

jtibshirani commented:
This looks good to me too 👍 We can keep an eye on the nightly benchmarks to double-check there's no change.

It's not directly related to this PR, but maybe we want to add an item to the meta issue under 'tech debt' about unifying the docvalues vs. VectorValues code paths (https://github.com/elastic/elasticsearch/issues/84324)? This could help cut down on the number of classes and code complexity.

jdconrad commented:
@jtibshirani Thanks for reviews as well!

jdconrad committed:
Merge branch 'main' into 8bitfd4

jdconrad commented:
Looks like the bwc failures in 7.1.1 and earlier are a known issue (https://github.com/elastic/elasticsearch/issues/91470), so I'm going to go ahead and push this change.

benwtrent commented:
Such a great feature @jdconrad!!! Integrating `byte` sized vectors us definitely a release highlight for 8.6.0!!!!

elasticsearchmachine commented:
@jdconrad according to this PR's labels, I need to update the changelog YAML, but I can't because the PR is closed. Please either update the changelog yourself on the appropriate branch, or adjust the labels. Specifically:

   * The PR is labelled `release highlight` but the changelog has no `highlight` section
